zebra/zebra_ns.c

```diff
@@ -57,10 +74,79 @@ int zebra_ns_enable(ns_id_t ns_id, void **info)
 	return 0;
 }
 
+struct route_table *zebra_ns_find_table(struct zebra_ns *zns,
+					struct zebra_vrf *zvrf,
+					uint32_t tableid, afi_t afi)
+{
+	struct zebra_ns_tables finder;
+	struct zebra_ns_tables *znst;
+
+	memset(&finder, 0, sizeof(finder));
+	finder.afi = afi;
+	finder.tableid = tableid;
+	znst = RB_FIND(zebra_ns_tables_head, &zns->ns_tables, &finder);
+
+	if (znst)
+		return znst->table;
+	else
+		return NULL;
+}
+
+struct route_table *zebra_ns_get_table(struct zebra_ns *zns,
+				       struct zebra_vrf *zvrf, uint32_t tableid,
+				       afi_t afi)
+{
+	struct zebra_ns_tables finder;
+	struct zebra_ns_tables *znst;
+	rib_table_info_t *info;
+
+	memset(&finder, 0, sizeof(finder));
+	finder.afi = afi;
+	finder.tableid = tableid;
+	znst = RB_FIND(zebra_ns_tables_head, &zns->ns_tables, &finder);
+
+	if (znst)
+		return znst->table;
+
+	znst = XCALLOC(MTYPE_ZEBRA_NS, sizeof(*znst));
+	znst->tableid = tableid;
+	znst->afi = afi;
+	znst->table =
+		(afi == AFI_IP6) ? srcdest_table_init() : route_table_init();
+
+	info = XCALLOC(MTYPE_RIB_TABLE_INFO, sizeof(*info));
+	info->zvrf = zvrf;
+	info->afi = afi;
+	info->safi = SAFI_UNICAST;
+	znst->table->info = info;
+	znst->table->cleanup = zebra_rtable_node_cleanup;
+
+	RB_INSERT(zebra_ns_tables_head, &zns->ns_tables, znst);
+	return znst->table;
+}
+
+static struct zebra_ns_tables *zebra_ns_free_table(struct zebra_ns_tables *znst)
+{
+	void *table_info;
+	rib_close_table(znst->table);
+
+	table_info = znst->table->info;
+	route_table_finish(znst->table);
+	XFREE(MTYPE_RIB_TABLE_INFO, table_info);
+	XFREE(MTYPE_ZEBRA_NS, znst);
+	return NULL;
+}
+
 int zebra_ns_disable(ns_id_t ns_id, void **info)
 {
+	struct zebra_ns_tables *znst;
 	struct zebra_ns *zns = (struct zebra_ns *)(*info);
 
+	while ((znst = RB_ROOT(zebra_ns_tables_head, &zns->ns_tables))
```

Comment:
`checkpatch.pl` will complain about assignments in conditionals.

Reply:
<sigh> I was following the code structures already in the tree for doing this.  Let's get together with @rwestphal and figure out the patter we want

Reply:
I agreewith Quentin. This will not pass checkstyle.
we should split.